### PR TITLE
[ASTGen] Syntax-tree based interface hash

### PR DESCRIPTION
**Pull request by rintaro**
Added `StableHasher` in ASTGen. Mostly copied from `stdlib/public/core/Hasher.swift` and ``stdlib/public/core/SipHash.swift``

For now, interface hash / fingerprint are emitted during the AST generation, but it should really be a separate request. I will separate fingerprint things from `ParseSourceFileRequest` and `ParseMembersRequest` in followups

https://github.com/swiftlang/swift/issues/76527
rdar://117150870

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/ASTGen/Sources/ASTGen/StableHasher.swift [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+
+extension StableHasher {
+  struct _State {
+    // "somepseudorandomlygeneratedbytes"
+    private var v0: UInt64 = 0x736f6d6570736575
+    private var v1: UInt64 = 0x646f72616e646f6d
+    private var v2: UInt64 = 0x6c7967656e657261
+    private var v3: UInt64 = 0x7465646279746573
+    // The fields below are reserved for future use. They aren't currently used.
+    private var v4: UInt64 = 0
```
Comment (CodaFi) [inline lib/ASTGen/Sources/ASTGen/StableHasher.swift:231]: I think we can delete these.

**Review by hamishknight — APPROVED** [review]

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test Linux

**Pull request merged**